sfs/time/soundfield.py

```diff
@@ -1,48 +1,57 @@
 """Compute sound field."""
 
+from __future__ import division
 import numpy as np
-
 from .. import util
-from . import source
 from .. import defs
-
-
-def synthesize_p(d_line, channel_weight, x0, grid, t, fs=None):
-    """Compute sound field for given time sample.
-
-    D_line(m x n) specifies the signals m fed into each secondary source n.
-    Channel_weight is applied during integration, e.g. source tapering.
-    x0 are available secandary sources.
+from .source import point
+
+
+def p_array(x0, d, channel_weights, t, grid, source=point, fs=None, c=None):
+    """Compute sound field for an array of secondary sources for a given time.
+
+    Parameters
+    ----------
+    x0 : (N, 3) array_like
+        Sequence of secondary source positions.
+    d : (N, C) array_like
+        Specifies the signals (with N samples) fed into each secondary source
+        channel C (columns).
+    channel_weights : (C,) array_like
+        Additional weights applied during integration, e.g. source tapering.
+    t : float
+        Simulation point in time (seconds).
+    grid : triple of array_like
+        The grid that is used for the sound field calculations.
+        See `sfs.util.xyz_grid()`.
+    source: function, optional
+        Source type is a function, returning scalar field.
+        For default, see `sfs.time.source.point()`.
+    fs: int, optional
+        Sampling frequency in Hertz.
+    c : float, optional
+        Speed of sound.
+
+    Returns
+    -------
+    numpy.ndarray
+        Sound pressure at grid positions.
 
     """
     if fs is None:
         fs = defs.fs
-    channel_weight = util.asarray_1d(channel_weight)
-    if len(channel_weight) != len(x0):
-        raise ValueError("length mismatch")
-
-    delay = np.nonzero(d_line)[0]
-    offset = np.min(delay)
-    d_line = d_line[offset:, :]  # remove driving function time offset
-
-    channels = len(x0)  # extract no. of channels
-    # Add additional zeros to the driving signal to ensure an amplitude of 0
-    # in the whole listening area before and after the real driving signal
-    padding_samples = 1000  # TODO: round(max(max_distance/c*fs,2*L/c*fs))
-    zero_fill = np.zeros([padding_samples, channels])
-
-    # time reversal driving function
-    sig_length = len(d_line)
-    t_inverted = t - sig_length - padding_samples  # compensate time
-    # zero padding and reverse time signals
-    d_line_inverted = np.flipud(np.row_stack((zero_fill, d_line, zero_fill)))
-
+    if c is None:
+        c = defs.c
+    x0 = util.asarray_of_rows(x0)
+    channel_weights = util.asarray_1d(channel_weights)
+    d = np.asarray(d)
+    if (len(channel_weights) != x0.shape[0]) or (
+            len(channel_weights) != d.shape[1]):
```

Comment:
No need for `or` here, you can just chain the comparisons. Yay Python!

And `x0.shape[0]` is equivalent to `len(x0)`. I would prefer latter.